lib/scholar/impute/simple_imputer.ex

```diff
@@ -0,0 +1,242 @@
+
+    * `:statistics` - The imputation fill value for each feature. Computing statistics can result in `Nx.Constant.nan/1` values.
+
+  ## Examples
+
+    iex> x = Nx.tensor([[1, 2, :nan], [3, 7, :nan], [:nan, 4, 5]])
+    iex> imputer = Scholar.Impute.SimpleImputer.fit(x, strategy: :mean)
+    iex> Scholar.Impute.SimpleImputer.transform(imputer, x)
+    #Nx.Tensor<
+      f32[3][3]
+      [
+        [1.0, 2.0, 5.0],
+        [3.0, 7.0, 5.0],
+        [2.0, 4.0, 5.0]
+      ]
+    >
+
+    iex> x = Nx.tensor([[1, 2, :nan], [3, 7, :nan], [:nan, 4, 5]])
+    iex> y = Nx.tensor([[7, :nan, 6], [6, 9, :nan], [8, :nan, 1]])
+    iex> imputer = Scholar.Impute.SimpleImputer.fit(x, strategy: :median)
+    iex> Scholar.Impute.SimpleImputer.transform(imputer, y)
+    #Nx.Tensor<
+      f32[3][3]
+      [
+        [7.0, 4.0, 6.0],
+        [6.0, 9.0, 5.0],
+        [8.0, 4.0, 1.0]
+      ]
+    >
+
+  """
+  deftransform fit(x, opts \\ []) do
+    validated_opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    input_rank = Nx.rank(x)
+
+    if input_rank != 2 do
+      raise ArgumentError, "Wrong input rank. Expected: 2, got: #{inspect(input_rank)}"
+    end
+
+    if validated_opts[:missing_values] != :nan and
+         Nx.any(Nx.is_nan(x)) == Nx.tensor(1, type: :u8) do
+      raise ArgumentError,
+            ":missing_values other than :nan possible only if there is no Nx.Constant.nan() in the array"
+    end
+
+    {type, _num_bits} = Nx.type(x)
+
+    x =
+      if validated_opts[:strategy] == :constant do
+        cond do
+          is_float(validated_opts[:fill_value]) and type in [:s, :u] ->
+            Nx.as_type(x, :f32)
+
+          is_integer(validated_opts[:fill_value]) and type in [:f, :bf] ->
+            {fill_value_type, _} = Nx.type(validated_opts[:fill_value])
+
+            raise ArgumentError,
+                  "Wrong type of `:fill_value` for the given data. Expected: :f or :bf, got: #{inspect(fill_value_type)}"
+
+          true ->
+            x
+        end
+      else
+        x
+      end
+
+    x_type = Nx.type(x)
+
+    x =
+      if validated_opts[:missing_values] != :nan,
+        do: Nx.select(Nx.equal(x, validated_opts[:missing_values]), Nx.Constants.nan(), x),
+        else: x
+
+    {_num_rows, num_cols} = Nx.shape(x)
+
+    statistics =
+      cond do
+        opts[:strategy] == :mean ->
+          mean_op(x)
+
+        opts[:strategy] == :median ->
+          median_op(x)
+
+        opts[:strategy] == :mode ->
+          mode_op(x, type: x_type)
+
+        true ->
+          Nx.broadcast(validated_opts[:fill_value], {num_cols})
+      end
+
+    missing_values = validated_opts[:missing_values]
+    %__MODULE__{statistics: statistics, missing_values: missing_values}
+  end
+
+  defnp mean_op(x) do
+    mask = not Nx.is_nan(x)
+    denominator = Nx.sum(mask, axes: [0])
+    temp = Nx.select(mask, x, 0)
+    nominator = Nx.sum(temp, axes: [0])
+
+    Nx.select(
+      denominator != 0,
+      nominator / denominator,
```

Comment:
This select should be over the denominator, replacing zeros with nans

Otherwise I think integer tensors can result in exceptions

Reply:
I think I don't understand. At this point, all tensors are float tensors since I made a NaN replacement

```elixir
x =
  if validated_opts[:missing_values] != :nan,
    do: Nx.select(Nx.equal(x, validated_opts[:missing_values]), Nx.Constants.nan(), x),
    else: x
```

Only the mode and constant imputation are reasonable for integer tensors to keep the type.